wagtail/wagtailadmin/static_src/wagtailadmin/js/modal-workflow.js

```diff
@@ -51,9 +51,15 @@ function ModalWorkflow(opts) {
     };
 
     self.loadResponseText = function(responseText) {
-        var response = eval('(' + responseText + ')');
+        var response;
 
-        self.loadBody(response);
+        try {
+            response = JSON.parse( responseText );
+        } catch ( error ) {
+            window.location = window.location.host;
```

Comment:
That's a fair point. If the user knows their session has timed out before submitting the form, they can log in in another tab, or copy their work into a text document... neither of which is particularly desirable. There's certainly a lot we could do to improve the user experience there, but that's out of scope for this issue. The important thing here is not to make the current experience worse (which would happen here, if opening a page chooser model caused their entire work to be lost).

I think the ideal response to opening a modal while logged out would be to display the login form within the modal, but that's probably a deep rabbit-hole to go down (to do it properly, you'd need to embed the 'forgotten password' workflow in the modal too, and handle non-standard login mechanisms like OAuth...). Failing that, a message in the modal saying "your session has timed out, please log in in another tab" would probably be the next best thing. (If we're going this route, it would be nice to have the server return a specific response that indicates 'session timeout', rather than just assuming that any non-JSON response is a session timeout - not essential, though.)